cyper-object-store/src/signer.rs

```diff
@@ -0,0 +1,207 @@
+    header_value.push('/');
+    header_value.push_str(region);
+    header_value.push('/');
+    header_value.push_str("s3");
+    header_value.push('/');
+    header_value.push_str("aws4_request");
+
+    // Signed headers
+    header_value.push_str(", SignedHeaders=");
+    header_value.push_str(signed_headers);
+
+    // Signature (HMAC of StringToSign)
+    header_value.push_str(", Signature=");
+    header_value.push_str(signature);
+
+    header_value
+}
+
+/// Generate the canonical string representation for an HTTP request, expected
+/// as the input to the AWS V4 signing algorithm.
+pub fn canonical_request(request: &Request) -> (String, String) {
+    let mut canonical_request = String::new();
+    // Method
+    canonical_request.push_str(request.method().as_str());
+    canonical_request.push('\n');
+
+    // Path
+    canonical_request.push_str(request.url().path());
+    canonical_request.push('\n');
+
+    // Query string
+    canonical_request.push_str(request.url().query().unwrap_or_default());
+    canonical_request.push('\n');
+
+    // Canonical headers
+    let mut canonical_headers = BTreeMap::new();
+    let mut signed_headers = BTreeSet::new();
+    for (name, value) in request.headers() {
+        if let Ok(value_str) = value.to_str() {
+            signed_headers.insert(canonical_header_name(name));
+            canonical_headers.insert(canonical_header_name(name), value_str.to_string());
+        }
+    }
+    for (name, value) in canonical_headers {
+        canonical_request.push_str(format!("{name}:{value}\n").as_str());
+    }
+    canonical_request.push('\n');
+
+    // Signed Headers
+    let signed_headers = signed_headers.iter().cloned().collect::<Vec<_>>().join(";");
+    canonical_request.push_str(signed_headers.as_str());
+    canonical_request.push('\n');
+
+    // TODO(aduffy): handle non-empty payloads
+    // Hash of payload bytes
+    canonical_request.push_str(EMPTY_CHECKSUM);
+
+    (canonical_request, signed_headers)
+}
+
+fn string_to_sign(canonical_request: &str, now: SystemTime, region: &str) -> String {
+    let mut to_sign = String::new();
+
+    to_sign.push_str("AWS4-HMAC-SHA256\n");
+
+    let timestamp: DateTime<Utc> = now.into();
+    to_sign.push_str(timestamp.format("%Y%m%dT%H%M%SZ").to_string().as_str());
+    to_sign.push('\n');
+
+    let scope = format!("{}/{region}/s3/aws4_request\n", timestamp.format("%Y%m%d"));
+    to_sign.push_str(scope.as_str());
+
+    let hash = Sha256::digest(canonical_request);
+    to_sign.push_str(hex::encode(hash).as_str());
+
+    to_sign
+}
+
+fn canonical_header_name(name: &HeaderName) -> String {
+    name.as_str().to_lowercase()
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::SystemTime;
+
+    use cyper::Client;
+    use http::header::{AUTHORIZATION, RANGE};
+    use http::HeaderValue;
+
+    use super::sign_request;
+
+    #[compio::test]
+    async fn test_signer() {
+        let client = Client::new();
+        let request = client
```

Comment:
so far i've tested this with S3 and the new Hetzner object storage, both range and non-range requests